src/lib/io/sol-uart-impl-linux.c

```diff
@@ -172,72 +271,72 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
 SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
+    struct sol_blob *blob;
+    uint16_t i;
+
     SOL_NULL_CHECK(uart);
-    if (uart->async.rx_fd_handler)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
+
+    if (uart->rx.used)
+        uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+
+    if (uart->fd_handler)
+        sol_fd_del(uart->fd_handler);
+
+    if (uart->read_timeout)
+        sol_timeout_del(uart->read_timeout);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i)
+        sol_blob_unref(blob);
+    sol_ptr_vector_clear(&uart->pending_blobs);
+    sol_buffer_fini(&uart->rx);
     close(uart->fd);
     free(uart);
 }
 
-static void
-uart_tx_dispatch(struct sol_uart *uart, int status)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    uart->async.tx_fd_handler = NULL;
-    if (!uart->async.tx_cb)
-        return;
-    uart->async.tx_cb((void *)uart->async.tx_user_data, uart,
-        (uint8_t *)uart->async.tx_buffer, status);
-}
+    int r;
+    size_t total;
+    bool fd_created = false;
 
-static bool
-uart_tx_callback(void *data, int fd, uint32_t active_flags)
-{
-    struct sol_uart *uart = data;
-    int ret;
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
 
-    if (active_flags & FD_ERROR_FLAGS) {
-        SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
-        ret = -1;
-        goto error;
-    }
+    r = sol_util_size_add(uart->pending_tx, blob->size, &total);
+    SOL_INT_CHECK(r, < 0, r);
 
-    if (uart->async.tx_index == uart->async.tx_length) {
-        uart_tx_dispatch(uart, uart->async.tx_index);
-        return false;
-    }
+    if (total >= uart->tx_size)
+        return -ENOSPC;
 
-    ret = write(fd, uart->async.tx_buffer + uart->async.tx_index,
-        uart->async.tx_length - uart->async.tx_index);
-    if (ret < 0) {
-        SOL_ERR("Error when writing to file descriptor %d.", fd);
-        goto error;
+    if (!uart->fd_handler) {
+        uart->fd_handler = sol_fd_add(uart->fd,
+            FD_ERROR_FLAGS | SOL_FD_FLAGS_OUT, uart_fd_handler_callback, uart);
+        SOL_NULL_CHECK(uart->fd_handler, -ENOMEM);
+        fd_created = true;
+    } else {
+        bool err;
+
+        err = sol_fd_add_flags(uart->fd_handler, SOL_FD_FLAGS_OUT);
```

Comment:
either here or in the `sol_fd_add_flags()` (https://github.com/solettaproject/soletta/blob/master/src/lib/common/sol-mainloop.c#L334) we'd need to check if the flag was already set, in that case avoid marking the "main loop" as dirty (ie: https://github.com/solettaproject/soletta/blob/master/src/lib/common/sol-mainloop-impl-posix.c#L753)

I guess on `sol-mainloop.c` it may benefit more people without adding complexity